tx/rm_stm: tighten lso computation and monotonicity

bharathv opened the pull request:
The main issue is if the stm is resetting state (at shutdown or applying snapshots), it may temporarily be in an inconsistent state (eg: active transactions list cleared) and a racy LSO request exactly at that point may result in overestimation of offset.  This was causing tx + compaction control batch removal test to flake during cross shard movement of replicas. 



## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.3.x
- [x] v25.2.x
- [x] v25.1.x
- [ ] v24.3.x

## Release Notes
* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

bharathv committed:
tx/rm_stm: do not compute lso at shutdown

As the stm is shutting down, the state the function is looking at may
be partial and there is a chance LSO is overestimated which can include
open transactions.

bharathv committed:
tx/rm_stm: misc comment edits

bharathv committed:
tx/rm_stm: do not compute LSO when reset is in progress

When a reset is in progress (eg: raft snapshot), the stm may clear all
the inflight transactions and reset the next apply offset. Due to
scheduling points, if an LSO request comes in racily, it may not see a
consistent state of the stm.

This commit returns the last known lso as a conservative approach until
the reset is finished.

bharathv committed:
tx/rm_stm/tests: add a test for monotonicity of lso

bashtanov commented on src/v/cluster/rm_stm.cc:
```diff
@@ -1850,14 +1850,15 @@ model::offset rm_stm::to_log_offset(kafka::offset k_offset) const {
       hdr.offset);
     tx_snapshot_v6 data;
-    iobuf_parser data_parser(std::move(tx_ss_buf));
+    iobuf_parser data_parser(std::move(data_buf));
```
probably belongs to a different commit

bharathv replied:
done.

WillemKauf commented on src/v/cluster/rm_stm.cc:
```diff
@@ -1264,6 +1264,7 @@ ss::future<result<kafka_result>> rm_stm::replicate_msg(
 }
 
 model::offset rm_stm::last_stable_offset() {
+    _as.check();
```
Is this exception caught appropriately in the respective call-stacks everywhere `last_stable_offset()` is used? I asked because previously this function didn't throw.

bharathv replied:
not really sure about this.. I handled this scenario gracefully by returning invalid_lso.

bharathv committed:
tx/rm_stm: avoid a dangling rvalue reference from function arg

Move it on to the stack. A deeper fix is to pass it by copy but thats a
bigger change and can be done later.

bharathv committed:
tx/rm_stm: avoid overlfow in timeout.

Using long_max milliseconds will result in overflow during internal
duration casts. instead use a large-ish timeout equivalent to no
timeout.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--bca03acc-4947-4218-a64d-58f29daf9ef4-->
<!--TEST RESULTS-->
<!--0b69c3dd-704c-453a-b1b8-cbd0de395350-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75587">build#75587</a></summary>


| test_class                      | test_method               | test_arguments                                                               | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                           |
|:--------------------------------|:--------------------------|:-----------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingReplicationTests   | test_auto_prefix_trimming | {"source_cluster_spec": {"cluster_type": "redpanda"}, "with_failures": true} | integration | https://buildkite.com/redpanda/redpanda/builds/75587#019a508d-08c9-41ed-9715-26a5a76f0550 | FLAKY         | 18/21    | upstream reliability is '92.41192411924119'. current run reliability is '85.71428571428571'. drift is 6.69764 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_auto_prefix_trimming  |
| RedpandaNodeOperationsSmokeTest | test_node_ops_smoke_test  | {"cloud_storage_type": 1, "mixed_versions": false}                           | integration | https://buildkite.com/redpanda/redpanda/builds/75587#019a508d-08c5-44f5-a59b-459dc68a525a | FLAKY         | 16/21    | upstream reliability is '100.0'. current run reliability is '76.19047619047619'. drift is 23.80952 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test |
| RedpandaNodeOperationsSmokeTest | test_node_ops_smoke_test  | {"cloud_storage_type": 1, "mixed_versions": true}                            | integration | https://buildkite.com/redpanda/redpanda/builds/75587#019a508d-08c6-4ec0-a24a-6b97b8c206ba | FLAKY         | 12/21    | upstream reliability is '99.21996879875195'. current run reliability is '57.14285714285714'. drift is 42.07711 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75625">build#75625</a></summary>


| test_class                      | test_method                                  | test_arguments                                                                                                      | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                             |
|:--------------------------------|:---------------------------------------------|:--------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ConsumerGroupBalancingTest      | test_coordinator_nodes_balance               | null                                                                                                                | integration | https://buildkite.com/redpanda/redpanda/builds/75625#019a5264-5c9f-429f-8d2c-3aeda76ef5d3 | FLAKY         | 20/21    | upstream reliability is '99.75186104218362'. current run reliability is '95.23809523809523'. drift is 4.51377 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ConsumerGroupBalancingTest&test_method=test_coordinator_nodes_balance  |
| MountUnmountIcebergTest         | test_simple_remount                          | {"cloud_storage_type": 1}                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/75625#019a5264-5c9f-429f-8d2c-3aeda76ef5d3 | FLAKY         | 18/21    | upstream reliability is '97.66702014846236'. current run reliability is '85.71428571428571'. drift is 11.95273 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=MountUnmountIcebergTest&test_method=test_simple_remount                |
| SegmentMsTest                   | test_segment_rolling_with_retention_consumer | null                                                                                                                | integration | https://buildkite.com/redpanda/redpanda/builds/75625#019a526b-4d11-4f1c-9672-793a456fd108 | FLAKY         | 19/21    | upstream reliability is '94.73039215686273'. current run reliability is '90.47619047619048'. drift is 4.2542 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SegmentMsTest&test_method=test_segment_rolling_with_retention_consumer |
| NodesDecommissioningTest        | test_recommissioning_node                    | {"cloud_topic": false}                                                                                              | integration | https://buildkite.com/redpanda/redpanda/builds/75625#019a5264-5c99-487a-baa8-27f8687c12fd | FLAKY         | 19/21    | upstream reliability is '97.64309764309765'. current run reliability is '90.47619047619048'. drift is 7.16691 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_recommissioning_node         |
| RedpandaNodeOperationsSmokeTest | test_node_ops_smoke_test                     | {"cloud_storage_type": 1, "mixed_versions": true}                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/75625#019a526b-4d1b-46ff-a6ef-54a0ffa819fe | FLAKY         | 10/21    | upstream reliability is '95.78947368421052'. current run reliability is '47.61904761904761'. drift is 48.17043 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test   |
| TieredStorageTest               | test_tiered_storage                          | {"cloud_storage_type_and_url_style": [1, "path"], "test_case": {"name": "(TS_Read == True, TS_Timequery == True)"}} | integration | https://buildkite.com/redpanda/redpanda/builds/75625#019a526b-4d16-4f50-a3d5-e7c86be3578a | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TieredStorageTest&test_method=test_tiered_storage                      |
</details>
<!--/body-->

bharathv committed:
maybe fix

bharathv committed:
tx/rm_stm: disable test_lso_bound_by_open_tx

bharathv commented:
force push fixed duration overflow due to ms::max() duration.

bashtanov commented on src/v/cluster/tests/rm_stm_tests.cc:
```diff
@@ -981,3 +981,89 @@ FIXTURE_TEST(test_concurrent_producer_evictions, rm_stm_test_fixture) {
+
+FIXTURE_TEST(test_lso_bound_by_open_tx, rm_stm_test_fixture) {
+    create_stm_and_start_raft();
+    auto& stm = *_stm;
+    auto raft = _raft;
```
is it to save 1 byte on each line that uses either of these?

bashtanov commented on src/v/cluster/tests/rm_stm_tests.cc:
```diff
@@ -981,3 +981,78 @@ FIXTURE_TEST(test_concurrent_producer_evictions, rm_stm_test_fixture) {
+          auto current_lso = stm.last_stable_offset();
+          BOOST_REQUIRE(current_lso >= last_lso);
+          last_lso = current_lso;
+          return ss::now();
```
This relies on `ss::do_until` injecting scheduling points even when the iteration is known to return readily available future. I'm not sure it's the case for both debug and release mode, and is going to remain like this. Maybe add `ss::maybe_yield` or another explicitly yielding call?

mmaslankaprv commented:
parts of this PR was already merged
